src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,365 @@
+#endif
+
+            // This could still be null if we had no match, but there's nothing left for us to try.
+            return retVal;
+        }
+
+        // Looks up a symbol by name (caller is responsible for mangling), and if it's found returns a delegate for it.
+        // Returns null if the symbol is not found.
+        private TDelegate CreateDelegateForSymbolByName<TDelegate>(string symbolName) where TDelegate : class
+        {
+            Debug.Assert(((RuntimeType)typeof(TDelegate)).IsDelegate());
+
+            IntPtr symbolAddress = GetProcAddress(_hInstance, symbolName);
+            return (symbolAddress != IntPtr.Zero)
+                ? Marshal.GetDelegateForFunctionPointer<TDelegate>(symbolAddress)
+                : null;
+        }
+
+        // Looks up a symbol by ordinal and if it's found returns a delegate for it.
+        // Returns null if the symbol is not found.
+        private TDelegate CreateDelegateForSymbolByOrdinal<TDelegate>(ushort ordinal) where TDelegate : class
+        {
+            Debug.Assert(((RuntimeType)typeof(TDelegate)).IsDelegate());
+
+            IntPtr symbolAddress = GetProcAddress(_hInstance, (IntPtr)ordinal);
+            return (symbolAddress != IntPtr.Zero)
+                ? Marshal.GetDelegateForFunctionPointer<TDelegate>(symbolAddress)
+                : null;
+        }
+
+        private static void ParseCharsetAndCallingConvention(Type delegateType, out CharSet charSet, out CallingConvention callingConvention)
+        {
+            Debug.Assert(delegateType is RuntimeType);
+            Debug.Assert(((RuntimeType)delegateType).IsDelegate());
+
+            charSet = default;
+            callingConvention = default;
+
+            var attr = delegateType.GetCustomAttribute<UnmanagedFunctionPointerAttribute>(inherit: false);
+            if (attr != null)
+            {
+                charSet = attr.CharSet;
+                callingConvention = attr.CallingConvention;
+            }
+
+            // If no character set or calling convention was specified (or if the delegate explicitly asked
+            // to use the OS default calling convention), fall back to the OS defaults. We need to check
+            // for default because the developer could have passed '0' for any enum value, and during marshaling
+            // the CLR treats this as "use defaults".
+
+            if (charSet == default)
+            {
+                charSet = FallbackCharSet;
+            }
+
+            if (callingConvention == default || callingConvention == CallingConvention.Winapi)
+            {
+                callingConvention = FallbackCallingConvention;
+            }
+        }
+
+        /// <summary>
+        /// Equivalent to <see cref="TryGetDelegate{TDelegate}(string, bool, out TDelegate)"/> where <em>exactSpelling</em> is <see langword="false"/>.
+        /// </summary>
+        public bool TryGetDelegate<TDelegate>(string name, out TDelegate result) where TDelegate : class
```

Comment:
Ideally we'd use `where TDelegate : delegate`, but that language feature doesn't exist :(. This is the next best constraint and matches the constraint in `Marshal.GetDelegateForFunctionPointer<TDelegate>(IntPtr)`.

Reply:
`where TDelegate : System.Delegate` is coming in C# 7.3 (as is `: System.Enum`, you just can't use the keywords, `delegate` and `enum`, as of the current iteration)

Reply:
See https://github.com/dotnet/roslyn/pull/24443

Reply:
Yeah, the PR to support the constraints looks about as good as done (and *consuming* the delegate/enum constraints has always been supported anyway).

Reply:
Do you know when that version of the compiler will be made available to the coreclr and corefx projects? If it's dropping imminently then we can probably get the change in time for 2.1.

Reply:
I believe @VSadov would know. It requires, at the very least, for the `features/constraints` branch to be merged back into `master` before it is available.